### PR TITLE
Organize imports collation

### PR DESCRIPTION
**Pull request by rbuckton**
After some internal discussion about #51916 and #52090, I took a brief look at providing more control over the sorting behavior for "Organize imports" by adding some additional configuration options to `UserPreferences`:

- `organizeImportsCollation` - Indicates whether to perform "ordinal" (binary) sorting (i.e., by the numeric value associated with each code point), or "unicode" collation (using the Unicode Collation Algorithm/`Intl.Collator`). Default: `"ordinal"`
- `organizeImportsLocale` - Overrides the locale used for collation. Specify `"auto"` to use the UI locale. Only applies to `organizeImportsCollation: "unicode"`. Default: `"en"`
- `organizeImportsNumericCollation` - Indicates whether to collate decimal numeric strings by their integer value (i.e., `"a1" < "a2" < "a100"`) instead of just their code points (i.e., `"a1" < "a100" < "a2"`). Only applies to `organizeImportsCollation: "unicode"`. Default: `true`
- `organizeImportsAccentCollation` - Indicates whether to compare characters with accents or other diacritic marks as unequal from the base character. Only applies to `organizeImportsCollation: "unicode"`. Default: `true`.
- `organizeImportsCaseFirst` - Either `"upper"` to collate upper-case characters before lower-case, `"lower"` to collate lower-case characters before upper-case, or `false` to collate using the default for the locale. Only applies to `organizeImportsCollation: "unicode"`. Default: `false`.

Each of these settings correlate to options passed to `Intl.Collator`. The `organizeImportsNumericCollation` and `organizeImportsCaseFirst` properties correspond to the `numeric` and `caseFirst` options to `Intl.Collator`. The combination of `organizeImportsIgnoreCase` and `organizeImportsAccentCollation` are used to determine the `sensitivity` of the collator:

| `organizeImportsIgnoreCase` | `organizeImportsAccentCollation` | `sensitivity` |
|:-:|:-:|:-:|
| `true` | `false` | `"base"` |
| `true` | `true` | `"accent"` |
| `false` | `false` | `"case"` |
| `false` | `true` | `"variant"` |

The default behavior is to perform an `"ordinal"` comparison in a way that is compatible with eslint's `sort-imports` rule, which is our current behavior. These collation rules only apply to import and export specifiers. The import statements themselves are sorted only by path, which continues to use `"ordinal"` sorting behavior.

These rules will also apply to auto-imports and import fixes, which will use the user's collation preferences to determine if a list of import or export specifiers is already sorted before selecting an insertion point.

These settings are not yet part of the VS Code settings schema, but can still be enabled using either the `typescript.unstable` or `javascript.unstable` configuration keys in settings.json:

```json
{
  "typescript.unstable": {
    "organizeImportsCollation": "unicode",
    "organizeImportsCaseFirst": "upper"
  }
}
```

### CODE REVIEW

**Comment by jakebailey** [comment]
The main thing I'll want to test is that it's possible to use sort-imports or simple-import-format (paired with some organize setting) and have organize be a no-op; that to me feels like the main blocker in that I know much of the team doesn't want to have ESLint enabled in the editor, so _something_ has to get things right enough to not require a `--fix` (in which case, may as well not even do anything).

I do wonder how much of this we can detect within the file; if I have a file sorted via one of these methods, it'd be cool to not have to configure anything and let it be auto. That would work in our codebase as only new/ambiguous files would conflict with a lint rule, which I think would be pretty rare.

**Comment by rbuckton** [comment]
> The main thing I'll want to test is that it's possible to use sort-imports or simple-import-format (paired with some organize setting) and have organize be a no-op; that to me feels like the main blocker in that I know much of the team doesn't want to have ESLint enabled in the editor, so _something_ has to get things right enough to not require a `--fix` (in which case, may as well not even do anything).

Ideally, the configuration options should be enough to match compatibility though I may have to extend it slightly to fall back to ordinal comparisons depending on sensitivity so as to match rules like `eslint-plugin-simple-import-sorting`.

The default behavior (`"ordinal"`) matches our current behavior in that it uses "eslint compatible" ordinal comparisons (where case-insensitivity/pseudo-case-folding is performed via `.toLower()` rather than `.toUpper()`).

> I do wonder how much of this we can detect within the file; if I have a file sorted via one of these methods, it'd be cool to not have to configure anything and let it be auto. That would work in our codebase as only new/ambiguous files would conflict with a lint rule, which I think would be pretty rare.

For now, I've chosen to mostly leave sort detection alone (in that it only tests for case sensitive vs. case insensitive), except that it now uses the collation settings provided. I did experiment with collation settings detection, but I felt it was a bit unreliable since there are a lot of moving parts (such as determining the collation locale).

**Comment by jakebailey** [comment]
I finally had a chance to test this; it looks like (probably intentionally), using your example config produces an output that matches simple-import-sort (besides separating out blocks), which is awesome!

I've also tested this PR with #52090 applied to #51590 and it also works as expected. Very cool.

**Comment by jakebailey** [comment]
Hm, here's one case it differs. Organize says:

```
import {
    Baseline,
    Compiler,
    FileBasedTest,
    FileBasedTestConfiguration,
    getFileBasedTestConfigurationDescription,
    getFileBasedTestConfigurations,
    IO,
    RunnerBase,
    TestCaseParser,
    TestRunnerKind,
} from "./_namespaces/Harness";
import * as Utils from "./_namespaces/Utils";
import * as compiler from "./_namespaces/compiler";
import * as ts from "./_namespaces/ts";
import * as vpath from "./_namespaces/vpath";
```

But simple-import-sort says:

```
import * as compiler from "./_namespaces/compiler";
import {
    Baseline,
    Compiler,
    FileBasedTest,
    FileBasedTestConfiguration,
    getFileBasedTestConfigurationDescription,
    getFileBasedTestConfigurations,
    IO,
    RunnerBase,
    TestCaseParser,
    TestRunnerKind,
} from "./_namespaces/Harness";
import * as ts from "./_namespaces/ts";
import * as Utils from "./_namespaces/Utils";
import * as vpath from "./_namespaces/vpath";
```

which seems correct. It doesn't seem like this PR is applying the rules to the paths?

**Comment by rbuckton** [comment]
> [...] It doesn't seem like this PR is applying the rules to the paths?

No, it isn't.

**Comment by rbuckton** [comment]
I've updated the PR to apply collation settings to paths as well.

**Review by andrewbranch — APPROVED** [review]
One option for exposing these settings in VS Code could be to make a single dropdown of “presets” that match popular formatters / lint rules 🤷

File: tests/baselines/reference/api/tsserverlibrary.d.ts [hunk 1]
```diff
@@ -2802,7 +2802,58 @@ declare namespace ts {
...
+                 * This preference is ignored if {@link organizeImportsCollation} is not `"natural"`.
+                 *
+                 * Default: `"en"`
+                 */
+                readonly organizeImportsCollationLocale?: string;
+                /**
+                 * Indicates whether numeric collation should be used for digit sequences in strings. When `true`, will collate
+                 * strings such that `a1z < a2z < a100z`. When `false`, will collate strings such that `a1z < a100z < a2z`.
+                 *
+                 * This preference is ignored if {@link organizeImportsCollation} is not `"natural"`.
```
Comment (jakebailey) [inline tests/baselines/reference/api/tsserverlibrary.d.ts:2833]: This line mentions "natural" as an option for `organizeImportsCollation`, but it seems to only accept `"ordinal" | "unicode"`

File: tests/baselines/reference/api/tsserverlibrary.d.ts [hunk 2]
```diff
@@ -2802,7 +2802,58 @@ declare namespace ts {
...
+                 * Indicates whether imports should be organized in a case-insensitive manner.
+                 */
                 readonly organizeImportsIgnoreCase?: "auto" | boolean;
+                /**
+                 * Indicates whether imports should be organized via an "ordinal" (binary) comparison using the numeric value
+                 * of their code points, or via "unicode" collation (via the
+                 * [Unicode Collation Algorithm](https://unicode.org/reports/tr10/#Scope)) using rules associated with the locale
+                 * specified in {@link organizeImportsCollationLocale}.
+                 *
+                 * NOTE: When comparing paths `"ordinal"` collation is always used.
```
Comment (jakebailey) [inline tests/baselines/reference/api/tsserverlibrary.d.ts:2815]: Is this still accurate after https://github.com/microsoft/TypeScript/pull/52115/commits/72678a3dde6be15abe5c583b6e769dceac2073e0 ? (I can't really tell)

Comment (rbuckton) [reply tests/baselines/reference/api/tsserverlibrary.d.ts:2815]: No, I'll update.

File: tests/baselines/reference/api/tsserverlibrary.d.ts (hunk 1 above)
Comment (rbuckton) [reply tests/baselines/reference/api/tsserverlibrary.d.ts:2833]: I switched from `"natural"` to `"unicode"`, but didn't update the comments. I'm looking at it now.

**Comment by rbuckton** [comment]
I started looking at detection of collation settings, in the same vein as `detectSortCaseSensitivity`, but I may wait on applying that.

**Comment by rbuckton** [comment]
> Hm, here's one case it differs. Organize says: [...]

@jakebailey, can you check if the current version now aligns with `simple-import-sort`?

**Comment by jakebailey** [comment]
That example appears to be fixed, yeah, thanks! One thing I did notice is that in order to get numeric sorting to match (sort `transformES5` before `transformES2015`), I also had to explicitly set:

```
"organizeImportsNumericCollation": true
```

But the doc comment seems to imply that it should be the default? (Maybe false is the default and the comment is outdated, or vice versa?)

File: src/services/organizeImports.ts [hunk 3]
```diff
@@ -699,3 +764,51 @@ function tryGetNamedBindingElements(namedImport: ImportDeclaration) {
...
 }
+
+function getOrganizeImportsOrdinalStringComparer(ignoreCase: boolean) {
+    return ignoreCase ? compareStringsCaseInsensitiveEslintCompatible : compareStringsCaseSensitive;
+}
+
+function getOrganizeImportsUnicodeStringComparer(ignoreCase: boolean, preferences: UserPreferences): Comparer<string> {
+    const resolvedLocale = getOrganizeImportsLocale(preferences);
+    const caseFirst = preferences.organizeImportsCaseFirst ?? false;
+    const numeric = preferences.organizeImportsNumericCollation ?? false;
```
Comment (jakebailey) [inline src/services/organizeImports.ts:775]: Ah, yeah, here's the mismatch; the doc comment says `true` is the default, but it's `false` here.

Comment (jakebailey) [reply src/services/organizeImports.ts:775]: I can't say I know what the default _should_ be; I don't think we did this previously so false is probably safer (though reasonably, I think true is "better", but maybe it's just "different").

Comment (rbuckton) [reply src/services/organizeImports.ts:775]: The default for `Intl.Collator` is `false`. I personally thing numeric collation is better, but wanted the default behavior for `organizeImpotsCollation` to match `new Intl.Collator(locale)`.

Comment (jakebailey) [reply src/services/organizeImports.ts:775]: Thanks for fixing the comment.

Comment (rbuckton) [reply src/services/organizeImports.ts:775]: Also, the default for `simple-import-sort` is apparently `{ sensitivity: "base", numeric: true }`, which means you'd need to set the following to match reliably:

```
{
  organizeImportsCollation: "unicode",
  organizeImportsIgnoreCase: true,
  organizeImportsNumericCollation: true,
}
```

Comment (jakebailey) [reply src/services/organizeImports.ts:775]: That's what it does as a first step, but if that comparison returns equal, it falls back to `a < b` i.e. ordinal (see https://github.com/lydell/eslint-plugin-simple-import-sort/#sorting)

These are the settings I've been using that seem to work:

```js
    "typescript.unstable": {
        "organizeImportsCollation": "unicode",
        "organizeImportsCaseFirst": "upper",
        "organizeImportsNumericCollation": true,
    },
```

**Comment by jakebailey** [comment]
Hm, just retested with:

```
    "typescript.unstable": {
        "organizeImportsCollation": "unicode",
        "organizeImportsCaseFirst": "upper",
        "organizeImportsNumericCollation": true,
    },
```

And I'm having trouble getting this to organize:


```
import {
    hasDecorators,
    HasDecorators,
} ...
```

`HasDecorators` should come first (to match simple-import-sort), but organize seems to accept it as sorted.

**Review by jakebailey — APPROVED** [review]

**Comment by jakebailey** [comment]
Just to finish this off, these are the settings now to make organize match simple-import-format:

```json
    "typescript.unstable": {
        "organizeImportsCollation": "unicode",
        "organizeImportsCaseFirst": "upper",
        "organizeImportsIgnoreCase": false,
        "organizeImportsNumericCollation": true,
    },
```

**Comment by rbuckton** [comment]
> [...] I'm having trouble getting this to organize [...]

Discussed offline and the issue was the default setting for `organizeImportsIgnoreCase` is `"auto"`, which auto-detects case sensitivity from context. Since `hasDecorators, HasDecorators` is a valid case-insensitive sort order, it acted as if `organizeImportsIgnoreCase` was `true`. The `organizeImportsCaseFirst` setting is only respected when perform a case-sensitive sort.

**Comment by jakebailey** [comment]
I believe that you're describing https://github.com/lydell/eslint-plugin-simple-import-sort/issues/127#issuecomment-1411612580.

Our general goal is to get the configurability (and behavior of various tools) to the point that we can offer up some presets that do "the right thing" for that plugin, dprint, etc, since this impacts auto-imports and it's nice to not have to immediately quick fix an eslint error.

**Comment by elboletaire** [comment]
I was about to open a bug as a regression of #23366, but after searching quite a lot I've found out these pull requests with these secret settings, and setting `organizeImportsIgnoreCase` to `true` seems to do what I want (note: using auto doesn't seem to be working well... since leaving it to "auto" was changing my imports from ignoring case to not ignoring it).

**Comment by mbest** [comment]
I'm seeing similar behavior. The detection is very inconsistent and seems to default case sensitive. In version 4.x, organize imports always used case-insensitive sorting (see #23366).

**Pull request merged**